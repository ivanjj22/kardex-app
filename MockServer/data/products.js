let products = [
  {
    id: "a775c47c-9961-41db-8275-3ae9dfbg0e04",
    name: "What is Lorem",
    description: "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
    img: "https://i.pcmag.com/imagery/reviews/046PYk5IcEZEDPsi0jh3hOE-27.1569480409.fit_scale.size_1028x578.jpg",
    price: 8000000,
    stock: 0,
    balance: 0
  },
  {
    id: "a775c47c-9961-dds32-8275-rwrd3434324",
    name: "Why do we use",
    description: "It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using 'Content here, content here', making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for 'lorem ipsum' will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).",
    img: "https://cdn.vox-cdn.com/thumbor/qIZjkb-fY8s2kQF05juGHmM46B4=/0x0:2040x1360/1200x800/filters:focal(857x517:1183x843)/cdn.vox-cdn.com/uploads/chorus_image/image/66714038/akrales_190913_3666_0391.0.jpg",
    price: 30000000,
    stock: 0,
    balance: 0
  },
];

let current_id = 1;


module.exports = {
  products,
  current_id
};
