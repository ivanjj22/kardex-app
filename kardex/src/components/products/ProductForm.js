import React from "react";

export const ProductForm = ({ onChange, onSubmit, form }) => {

  return (
    <form className="w-full" onSubmit={onSubmit}>

      <div className="flex flex-wrap -mx-3 ">
        <div className="w-full md:w-1/2 px-3 mb-6 md:mb-0">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="name"
          >
            Nombre
          </label>
          <input
            className="nameInput appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"
            name="name"
            type="text"
            placeholder="Name"
            onChange={onChange}
            value={form.name}
          />
        </div>
        <div className="w-full md:w-1/2 px-3">
          <label
            className="imageInput block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="img"
          >
            Imagen url
          </label>
          <input
            className="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="img"
            type="text"
            placeholder="Url"
            onChange={onChange}
            value={form.img}
          />
        </div>
      </div>
      <div className="flex flex-wrap -mx-3 mb-2">
        <div className="w-full  px-3 mb-6 md:mb-0">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="price"
          >
            Precio unitario
          </label>
          <input
            className="priceInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="price"
            type="number"
            placeholder="Precio"
            onChange={onChange}
            value={form.price}
          />
        </div>

        
        {form.id ? <div className="w-full px-3 mb-6 md:mb-0">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="price"
          >
            Stock
          </label>
          <input
            className="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="stock"
            type="number"
            placeholder="stock"
            onChange={onChange}
            value={form.stock}
            disabled={form.id}
          />
        </div>
        : null 
        }

      </div>
      <div className="flex flex-wrap -mx-3 mb-6">
        <div className="w-full px-3">
          <label
            className="descriptionInput block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="description"
          >
            Descripcion
          </label>
          <textarea
            className="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="description"
            rows="5"
            placeholder="Type a description"
            onChange={onChange}
            value={form.description}
          ></textarea>
        </div>
      </div>
      

      <div className="text-center mt-4">
        <button type="submit" className="btn btn-success text-white">
          Save
        </button>
      </div>
    </form>
  );
};
