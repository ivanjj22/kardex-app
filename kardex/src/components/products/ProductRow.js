import React from 'react'

import {
    Link
} from "react-router-dom";

export const ProductRow = ({product, index}) => {
    return (
        <tr key={product.id}>
            <th className="border px-8 py-4">{index}</th>
            <td className="border px-8 py-4">

                <div className="flex items-center space-x-3">
                <div className="avatar">
                    <div className="w-12 h-12 mask mask-squircle">
                    <img src={product.img} alt="Avatar Tailwind CSS Component"/>
                    </div>
                </div> 
                <div>
                    <div className="font-bold">
                        {product.name}
                        </div> 
                    <div className="text-sm opacity-50">
                        
                    </div>
                </div>
                </div>

            </td>
            <td className="border px-8 py-4">{product.description}</td>
            <td className="border px-8 py-4">{product.price}</td>
            <td className="border px-8 py-4 ">
                <Link to={`/products/${product.id}`} className="btn btn-primary btn-sm">
                    <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6 text-center cursor-pointer" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M15 12a3 3 0 11-6 0 3 3 0 016 0z" />
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M2.458 12C3.732 7.943 7.523 5 12 5c4.478 0 8.268 2.943 9.542 7-1.274 4.057-5.064 7-9.542 7-4.477 0-8.268-2.943-9.542-7z" />
                    </svg>
                </Link>
            </td>
        </tr>
    )
}
