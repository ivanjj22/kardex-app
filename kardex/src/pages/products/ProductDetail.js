import React, {useState} from "react";
import { Link } from 'react-router-dom'
import { useParams, Redirect } from "react-router-dom";
import useFetch from 'use-http'
import url from '../../config'
import {KardexForm} from '../../components/kardex/KardexForm'
import { KardexTable } from "../../components/kardex/KardexTable";
import { Loading } from "../../components/ui/Loading";


export const ProductDetail = () => {
  let { id } = useParams();

  const { data: { data:product } = {}, loading } = useFetch(`${url}/products/${id}`, {
    cachePolicy: 'no-cache'
  }, []);

  const [current, setCurrent] = useState(0);

  const onInsertKardex = () => {
    setCurrent(current+1);
  }

  if(loading) return <Loading/>

  if ( !product ) {
    return <Redirect to="/" />;
  }

  return (
    <>
        <div className="flex product-detail">
            <div className="flex-1 ">
                <div className="card text-center shadow-2xl">
                    <figure className="max-w-screen-lg mx-auto pb-3">
                        <img alt="Producto" style={{width: 50 + '%'}} src={product.img} className="w-12 rounded-xl mx-auto"/>
                    </figure> 
                    <div className="card-body">
                        <h2 className="card-title">{product.name}</h2> 
                        <p>{product.description}</p> 
                        <div className="justify-center card-actions">
                        <Link to={`/products/${id}/edit`} className="btn btn-outline btn-accent">Edit</Link>
                        </div>
                    </div>
                </div> 
            </div>
        </div>

        <div className="flex mt-5 ">
            <div className="flex-1">
                <h3 className="text-center text-3xl">Kardex</h3>

                <KardexForm onInsertKardex={onInsertKardex}/>
                
                

            </div>
        </div>

        <div className="flex mb-5">
            <div className="flex-1">
                
                <KardexTable current={current}/>

            </div>
        </div>

    </>
  );


};
