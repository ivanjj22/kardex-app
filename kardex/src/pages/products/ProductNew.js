import React from 'react'
import { ProductForm } from '../../components/products/ProductForm'

export const ProductNew = ({form, onChange, onSubmit}) => {
    return (
        <div>
            <ProductForm
            onChange={onChange}
            onSubmit={onSubmit}
            form={form}
            />
        </div>
    )
}
