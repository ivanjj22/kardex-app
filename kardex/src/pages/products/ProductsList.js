import React from "react";
import { Link } from 'react-router-dom'
import useFetch from 'use-http'
import url from '../../config'
import { ProductRow } from "../../components/products/ProductRow";
import { Loading } from "../../components/ui/Loading";


export const ProductsList = () => {

  const { loading, data: {data} = [] } = useFetch(`${url}/products`, {
    cachePolicy: 'no-cache'
  }, [])

  if (!data) return null;

  if(loading) return <Loading/>

  return (
    <div className="overflow-x-auto">

      <Link to="/products/new" className="btn btn-success mt-2 mb-2 text-white">
          New
      </Link>

      <table className="shadow-lg bg-white w-full">
        <thead>
          <tr>
            <th className="bg-blue-100 border text-left px-8 py-4"></th>
            <th className="bg-blue-100 border text-left px-8 py-4">Product</th>
            <th className="bg-blue-100 border text-left px-8 py-4">Description</th>
            <th className="bg-blue-100 border text-left px-8 py-4">
              Price
            </th>
            <th className="bg-blue-100 border text-left px-8 py-4">
              
            </th>
          </tr>
        </thead>
        <tbody>
          {
            data.map((product, index) => (
              <ProductRow key={product.id} index={index+1} product={product}/>
            ))
          }
        </tbody>
      </table>
    </div>
  );
};
