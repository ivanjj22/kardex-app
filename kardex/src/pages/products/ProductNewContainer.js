import React, { useState, useEffect } from "react";
import { ProductNew }  from "./ProductNew";
import url from "../../config";
import { useForm } from "../../hooks/useForm";
import { useParams } from "react-router-dom";
import useFetch from "use-http";

const ProductNewContainer = ({ history }) => {

  let { id } = useParams();

  const { get, post, put } = useFetch(`${url}`, {
    cachePolicy: 'no-cache'
  }, []);
  
  const [initialData, setInitialData] = useState({
    name: "",
    description: "",
    img: "",
    price: "",
    stock: 0
  });

  const [form, handleInputChange] = useForm(initialData, initialData);

  useEffect(() => {
    if (id) {
      loadItem();
    }
  }, [id]);

  const loadItem = async () => {
    const item = await get(`/products/${id}`);
    if (item.data) setInitialData(item.data);
  }

  const handleSubmit = async (e) => {
    e.preventDefault();
    if(!form.name || form.description || form.img) {
      return false;
    }
    try {
      if (id) {
        await put(`/products/${id}`, form);
      } else {
        await post(`/products`, form);
      }
      history.push("/products");
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <ProductNew form={form} onChange={handleInputChange} onSubmit={handleSubmit} />
  );
};

export default ProductNewContainer;
