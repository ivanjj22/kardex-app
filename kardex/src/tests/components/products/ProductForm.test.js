import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { ProductForm } from '../../../components/products/ProductForm';

describe('Pruebas en <ProductForm />', () => {

    test('debe de mostrarse correctamente', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/products/new']}>
                <Route 
                    path="/products/new" 
                    component={ () => <ProductForm form={{}} /> } 
                />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();

    })

    test('debe de hacer submit', () => {

        const history = {
            push: jest.fn()
        };
        const onChange = jest.fn();
        const onSubmit = jest.fn();

        const wrapper = mount(
            <MemoryRouter initialEntries={['/products/new']}>
                <Route 
                    path="/products/new" 
                    component={ () => <ProductForm history={ history } onSubmit={onSubmit} onChange={onChange} form={{}} /> } 
                />
            </MemoryRouter>
        );
        
        wrapper.find('.nameInput').simulate('change', {
            target: {
                name: 'name',
                value: 'Product test'
            }
        });

        wrapper.find('.imageInput').simulate('change', {
            target: {
                name: 'img',
                value: 'example.com'
            }
        });

        wrapper.find('.priceInput').simulate('change', {
            target: {
                name: 'price',
                value: '1000'
            }
        });

        wrapper.find('.descriptionInput').simulate('change', {
            target: {
                name: 'description',
                value: 'Description'
            }
        });
        

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        

        expect(onSubmit).toHaveBeenCalledTimes(1);
    })
    

    

    
})
