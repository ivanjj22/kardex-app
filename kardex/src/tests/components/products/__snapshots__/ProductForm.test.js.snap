// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ProductForm /> debe de mostrarse correctamente 1`] = `
<form
  className="w-full"
>
  <div
    className="flex flex-wrap -mx-3 "
  >
    <div
      className="w-full md:w-1/2 px-3 mb-6 md:mb-0"
    >
      <label
        className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="name"
      >
        Nombre
      </label>
      <input
        className="nameInput appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"
        name="name"
        placeholder="Name"
        type="text"
      />
    </div>
    <div
      className="w-full md:w-1/2 px-3"
    >
      <label
        className="imageInput block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="img"
      >
        Imagen url
      </label>
      <input
        className="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        name="img"
        placeholder="Url"
        type="text"
      />
    </div>
  </div>
  <div
    className="flex flex-wrap -mx-3 mb-2"
  >
    <div
      className="w-full  px-3 mb-6 md:mb-0"
    >
      <label
        className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="price"
      >
        Precio unitario
      </label>
      <input
        className="priceInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        name="price"
        placeholder="Precio"
        type="number"
      />
    </div>
  </div>
  <div
    className="flex flex-wrap -mx-3 mb-6"
  >
    <div
      className="w-full px-3"
    >
      <label
        className="descriptionInput block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="description"
      >
        Descripcion
      </label>
      <textarea
        className="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        name="description"
        placeholder="Type a description"
        rows="5"
      />
    </div>
  </div>
  <div
    className="text-center mt-4"
  >
    <button
      className="btn btn-success text-white"
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
